File: workspace/projects/heartwood-mcp/src/tools/jt/payments.ts
/**
 * JT Payment tools — 2 tools
 */

import type { PaveClient, ToolResult } from "../../pave/index.js";
import { PAYMENT_FIELDS } from "../../pave/index.js";
import type { ToolDef } from "../registry.js";
import { buildFilter, pickDefined, PAGINATION_PROPS, getPagination } from "./helpers.js";

export function paymentTools(pave: PaveClient): ToolDef[] {
  return [
    // ── jt_create_payment ──────────────────────────────────────────────
    {
      name: "jt_create_payment",
      description:
        "Create a payment on a document. Auto-detects credit/debit from document type.",
      inputSchema: {
        type: "object" as const,
        properties: {
          amount: { type: "number", description: "Payment amount" },
          date: { type: "string", description: "Payment date (YYYY-MM-DD)" },
          documentId: { type: "string", description: "Document ID" },
          description: { type: "string", description: "Payment description (optional)" },
          paymentType: { type: "string", description: "Payment type (optional)" },
        },
        required: ["amount", "date", "documentId"],
      },
      handler: async (params: Record<string, unknown>): Promise<ToolResult> => {
        const data: Record<string, unknown> = {
          amount: params.amount,
          date: params.date,
          documentId: params.documentId,
          ...pickDefined(params, ["description", "paymentType"]),
        };
        return pave.create("createPayment", data, PAYMENT_FIELDS);
      },
    },

    // ── jt_get_payments ────────────────────────────────────────────────
    {
      name: "jt_get_payments",
      description: "Get payments, optionally filtered by job, document, or account.",
      inputSchema: {
        type: "object" as const,
        properties: {
          jobId: { type: "string", description: "Filter by job ID (optional)" },
          documentId: { type: "string", description: "Filter by document ID (optional)" },
          accountId: { type: "string", description: "Filter by account ID (optional)" },
          ...PAGINATION_PROPS,
        },
        required: [],
      },
      handler: async (params: Record<string, unknown>): Promise<ToolResult> => {
        return pave.query({
          entityPlural: "payments",
          returnFields: PAYMENT_FIELDS,
          where: buildFilter(params, [
            { param: "jobId", field: "jobId" },
            { param: "documentId", field: "documentId" },
            { param: "accountId", field: "accountId" },
          ]),
          ...getPagination(params),
        });
      },
    },
  ];
}

File: heartwood-mcp/tools/payments.py
"""Payment tools — create and query payments."""

from __future__ import annotations

from app import mcp
from constants import ORG_ID
from pave import call_pave, flatten_nodes


@mcp.tool()
async def jt_create_payment(
    document_id: str,
    amount: float,
    payment_method: str | None = None,
    reference: str | None = None,
    date: str | None = None,
) -> dict:
    """Record a payment against a document (invoice).

    Args:
        document_id: The document (invoice) ID.
        amount: Payment amount.
        payment_method: Payment method (e.g. "check", "credit_card", "ach").
        reference: Reference number (check number, transaction ID, etc.).
        date: Payment date (YYYY-MM-DD). Defaults to today.
    """
    params: dict = {
        "organizationId": ORG_ID,
        "documentId": document_id,
        "amount": amount,
    }
    if payment_method:
        params["paymentMethod"] = payment_method
    if reference:
        params["reference"] = reference
    if date:
        params["date"] = date

    return await call_pave(
        "createPayment",
        params=params,
        fields={"id": {}, "amount": {}, "date": {}},
    )


@mcp.tool()
async def jt_get_payments(
    document_id: str | None = None,
    job_id: str | None = None,
    first: int = 50,
) -> list[dict]:
    """List payments, optionally filtered by document or job.

    Args:
        document_id: Filter by document (invoice).
        job_id: Filter by job.
        first: Max results.
    """
    filter_obj: dict = {}
    if document_id:
        filter_obj["documentId"] = {"eq": document_id}
    if job_id:
        filter_obj["jobId"] = {"eq": job_id}

    params: dict = {"first": first}
    if filter_obj:
        params["filter"] = filter_obj

    result = await call_pave(
        "payments",
        params=params,
        fields={
            "nodes": {
                "id": {},
                "amount": {},
                "date": {},
                "paymentMethod": {},
                "reference": {},
                "documentId": {},
            }
        },
    )
    return flatten_nodes(result)
